sim/midas/src/main/scala/midas/passes/fame/ImplementThreadedMems.scala

```diff
@@ -0,0 +1,172 @@
+// See LICENSE for license details.
+
+package midas.passes.fame
+
+import firrtl._
+import firrtl.ir._
+import firrtl.Utils.BoolType
+import firrtl.passes.MemPortUtils._
+
+import firrtl.Mappers._
+
+import collection.mutable
+
+/*
+ * The multithreaders add custom IR nodes to represent threaded memories.
+ * This pass finds them, replaces them with blackbox instances, and implements the blackboxes.
+ */
+
+case class ThreadedMem(nThreads: BigInt, proto: DefMemory) extends Statement with IsDeclaration {
+  require(proto.readLatency == 0 || proto.readLatency == 1)
+
+  val name = proto.name
+  val info = proto.info
+  def serialize = ???
+  def mapStmt(f: Statement => Statement): Statement = this
+  def mapExpr(f: Expression => Expression): Statement = this
+  def mapType(f: Type => Type): Statement = this.copy(proto = proto.copy(dataType = f(proto.dataType)))
+  def mapString(f: String => String): Statement = this.copy(proto = proto.copy(name = f(proto.name)))
+  def mapInfo(f: Info => Info): Statement = this.copy(proto = proto.copy(info = f(proto.info)))
+  def foreachExpr(f: Expression => Unit): Unit = ()
+  def foreachStmt(f: Statement => Unit): Unit = ()
+  def foreachType(f: Type => Unit): Unit = proto.foreachType(f)
+  def foreachString(f: String => Unit): Unit = proto.foreachString(f)
+  def foreachInfo(f: Info => Unit): Unit = proto.foreachInfo(f)
+}
+
+object ImplementThreadedMems {
+  private def wsf(e: Expression, f: String): WSubField = WSubField(e, f)
+  private def wssf(e: Expression, f0: String, f1: String): WSubField = WSubField(wsf(e, f0), f1)
+  private def prim(op: PrimOp, args: Seq[Expression]) = DoPrim(op, args, Nil, UnknownType)
+
+  private def implement(tMem: ThreadedMem, moduleName: String): Module = {
+    val info = FAME5Info.info
+    val tIdxMax = UIntLiteral(tMem.nThreads-1)
+    val bramToBufferPipeDepth = if (tMem.nThreads < 4) 0 else 1
```

Comment:
I know this existed previously, but it would be comment why this threshold is what it is.

Reply:
Good idea